framework/web/Request.php

```diff
@@ -817,6 +854,82 @@ public function getQueryParam($name, $defaultValue = null)
         return isset($params[$name]) ? $params[$name] : $defaultValue;
     }
 
+    /**
+     * Sets the data related to the incoming request environment.
+     * @param array $serverParams server parameters.
+     * @since 2.1.0
+     */
+    public function setServerParams(array $serverParams)
```

Comment:
There are `with*` methods for everything but `serverParams`. I think it makes sense to make it uniform having only `with*` public methods.

Reply:
All `with*` methods are dictated by PSR. Since PSR does not provide `withServerParams()` method, I see no reason for implementing it. If you find this inconsistent you should report it ot PHPFig.

Usage of `with*` methods is uneffective, since it causes object cloning. While configuring single `Request` instance via `with*` methods you will have it cloned 5-10 times.
Yii behaviors and filters would be unable to function via immutable setters as they may need to adjust passing `Request` or `Response` without being able to re-set thier origin.

Also keep in mind that while being cloned every instance of `yii\base\Component`, which `Request` is, looses any event handlers or behaviors (filters) attached to it.

Reply:
Also keep in mind that removing setters in favor of `with*` methods will make impossible to configure `Request` instance via DI, even with the recent features from `yiisoft/di`. For example:

```php
[
    '__class' => 'yii\web\Request',
    'withCookieParams' => [....], // have no effect: calls `$request->withCookieParams()`, but looses returned new instance
]
```

Reply:
The whole idea of the PSR is that Request object should be created once (passing everything needed in constructor) and then not being modified again except cloning via `with`.

Reply:
If you sure in which way it should be - feel free to make modifications.

Reply:
Do we have any request behaviors currently?

Reply:
Losing event handlers could be a problem. Need to check code again before I suggest anything...

Reply:
Currently Request doesn't trigger anything and there are no known behaviors for it.